### ISSUE TITLE
Checklist for ONNX operators vs. Fluid operators

### ISSUE THREAD

**Issue by lcy-seso** [PaddlePaddle/Paddle#8028]
The checklist is to check which operators Fluid has not supported yet according to the [operator list given by ONNX](https://github.com/onnx/onnx/blob/master/docs/Operators.md).

|ONNX|Fluid supports required elementary operator|Need enhancement|Need to wrap multiple elementary operators in Python|
|:--|:--:|:--:|:--:|
|Abs|:heavy_check_mark:|
|Add|:heavy_check_mark:
|[And](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Add)||NEED to support broadcast
|[ArgMax](https://github.com/onnx/onnx/blob/master/docs/Operators.md#argmax)|:x:|
|[ArgMin](https://github.com/onnx/onnx/blob/master/docs/Operators.md#argmax)|:x:|
|[AveragePool](https://github.com/onnx/onnx/blob/master/docs/Operators.md#AveragePool)|Need to support more pad mode.
|BatchNormalization|:heavy_check_mark:
|Cast|:heavy_check_mark:
|[Ceil](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Ceil)|:x:|
|[Clip](https://github.com/onnx/onnx/blob/master/docs/Operators.md#clip)|:heavy_check_mark:||
|Concat|:heavy_check_mark:
|Constant|||:heavy_check_mark:
|[Conv](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Conv)||Need to support more pad mode.
|ConvTranspose|:heavy_check_mark:
|DepthToSpace|:heavy_check_mark:
|Div|:heavy_check_mark:
|Dropout|:heavy_check_mark:
|Elu|:heavy_check_mark:
|Equal|:heavy_check_mark:
|Exp|:heavy_check_mark:
|Flatten|:heavy_check_mark:
|[Floor](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Floor)|:x:|
|GRU|:heavy_check_mark:
|[Gather](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Gather)|:x:|
|Gemm|:heavy_check_mark:
|GlobalAveragePool|:heavy_check_mark:
|GlobalLpPool|:heavy_check_mark:
|GlobalMaxPool|:heavy_check_mark:
|Greater|:heavy_check_mark:
|HardSigmoid|:heavy_check_mark:
|[Hardmax](https://github.com/onnx/onnx/blob/master/docs/Operators.md#hardmax)|:x:
|[InstanceNormalization](https://github.com/onnx/onnx/blob/master/docs/Operators.md#InstanceNormalization)|:x:|
|LRN|:heavy_check_mark:
|LSTM|:heavy_check_mark:
|LeakyRelu|:heavy_check_mark:
|Less|:heavy_check_mark:
|Log|:heavy_check_mark:
|LogSoftmax|||:heavy_check_mark:
|LpNormalization|:heavy_check_mark:
|LpPool|:heavy_check_mark:
|MatMul|:heavy_check_mark:
|Max|:heavy_check_mark:
|[MaxPool](https://github.com/onnx/onnx/blob/master/docs/Operators.md#MaxPool)||Need to support more pad mode.
|MaxRoiPool|:heavy_check_mark:
|Mean|:heavy_check_mark:
|Min|:heavy_check_mark:
|Mul|:heavy_check_mark:
|Neg|||:heavy_check_mark:
|[Not](https://github.com/onnx/onnx/blob/master/docs/Operators.md#not)||NEED to support broadcast
|[Or](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Or)||NEED to support broadcast
|PRelu|:heavy_check_mark:
|Pad|:heavy_check_mark:
|Pow|:heavy_check_mark:
|RNN|||:heavy_check_mark:
|RandomNormal|:heavy_check_mark:
|[RandomNormalLike](https://github.com/onnx/onnx/blob/master/docs/Operators.md#randomnormallike)|:x:
|[RandomUniform](https://github.com/onnx/onnx/blob/master/docs/Operators.md#RandomUniform)|:x:
|[RandomUniformLike](https://github.com/onnx/onnx/blob/master/docs/Operators.md#randomuniformlike)|:x:
|Reciprocal|:heavy_check_mark:
|ReduceL1|:heavy_check_mark:
|ReduceL2|:heavy_check_mark:
|ReduceLogSum|||:heavy_check_mark:
|ReduceLogSumExp|||:heavy_check_mark:
|ReduceMax|:heavy_check_mark:
|ReduceMean|:heavy_check_mark:
|ReduceMin|:heavy_check_mark:
|[ReduceProd](https://github.com/onnx/onnx/blob/master/docs/Operators.md#reduceprod)|:x:
|ReduceSum|:heavy_check_mark:
|ReduceSumSquare|||:heavy_check_mark:
|Relu|:heavy_check_mark:
|Reshape|:heavy_check_mark:
|[Selu](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Selu)|:x:
|Sigmoid|:heavy_check_mark:
|[Slice](https://github.com/onnx/onnx/blob/master/docs/Operators.md#slice)|:x:(under review)
|Softmax|:heavy_check_mark:
|Softplus|:heavy_check_mark:
|Softsign|:heavy_check_mark:
|SpaceToDepth|:heavy_check_mark:
|Split|:heavy_check_mark:
|Sqrt|:heavy_check_mark:
|[Squeeze](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Squeeze)|:x:
|Sub|:heavy_check_mark:
|Sum|:heavy_check_mark:
|Tanh|:heavy_check_mark:
|Tile|:heavy_check_mark:
|Transpose|:heavy_check_mark:
|[Unsqueeze](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Unsqueeze)|:x:|
|[Xor](https://github.com/onnx/onnx/blob/master/docs/Operators.md#xor)||Need to support broadcast|
|~~experimental ATen~~|
|experimental  Affine|||:heavy_check_mark:
|experimental ConstantFill|:heavy_check_mark:
|experimental Crop|:heavy_check_mark:
|experimental FC|:heavy_check_mark:
|experimental GRUUnit|:heavy_check_mark:
|experimental GivenTensorFill|:heavy_check_mark:
|experimental Identity|:heavy_check_mark:<br>assign_op in Fluid||
|[experimental ImageScaler](https://github.com/onnx/onnx/blob/master/docs/Operators.md#ImageScaler)|||:heavy_check_mark:|
|experimental MeanVarianceNormalization|:heavy_check_mark:
|experimental ParametricSoftplus|:x:|
|experimental Scale|:heavy_check_mark:
|experimental ScaledTanh|:x:
|experimental ThresholdedRelu|:heavy_check_mark:
|experimental Upsample||:heavy_check_mark:|

**Comment by varunarora** [comment]
@lcy-seso Can you give a little info on what it means to wrap multiple operators in Python? It isn't a requirement from the ONNX spec right?

**Comment by varunarora** [comment]
Second question: Do we have operators that are NOT in the ONNX spec? Which would require us to do some deprecated mapping?

**Comment by varunarora** [comment]
Extending the previous comment to types. Do we have a list of types that we need to deprecate? What functionality is absolutely not portable to ONNX? We know cycles, conditional blocks, and concurrency have some issues here right?

**Comment by varunarora** [comment]
Lastly, can we annotate the mappings with the Fluid name of the ops? This will help us better gauge the delta of work needed to be done

**Comment by lcy-seso** [comment]
Really sorry for the late reply.

1. for the first question: "wrap multiple operators in Python"

    Fluid does not directly implement operators like "[DepthToSpace](https://github.com/onnx/onnx/blob/master/docs/Operators.md#depthtospace)", "[SpaceToDepth](https://github.com/onnx/onnx/blob/master/docs/Operators.md#SpaceToDepth)", "[Squeeze](https://github.com/onnx/onnx/blob/master/docs/Operators.md#squeeze)", "[Unsqueeze](https://github.com/onnx/onnx/blob/master/docs/Operators.md#unsqueeze)" as standalone operators. These operators can be implemented by using "reshape_op" in Python end. Similar operators include [InstanceNorm](https://github.com/onnx/onnx/blob/master/docs/Operators.md#InstanceNormalization)

1. for the second question: Do we have operators that are NOT in the ONNX spec? 
    
    Yes, we do have many.
    - Especially, there is one particular category: the sequence processing related operators (for the reason that Fluid can support sequence processing in non-padding style, this is different from some other framework). like `sequence_expand`, `sequence_reshape`, `sequence_concat`, `sequence_pool` ...
    - The `while_loop` related operators, `reorder_lod_tensor_by_rank`, `shrink_rnn_memory`, ... 
    - The checklist above lists all the operators ONNX suggest to support, which does not include these Fluid's special operators. I guess we can move these special operators into our next step. We do not need a deprecated mapping at first.

1. Another problem is, I have not looked into the very details about how these special operators can be adapted to ONNX requirement. It seems that ONNX does not require to support these special operators, due to the different internal implementations, a potential problem may be some important model, for example, RNN encoder-decoder with attention model in NLP field, cannot be supported by only using ONNX operators. Wil this be a problem? Sorry, I am not sure about this.

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
